### PR TITLE
Add IDToken for Azure provider

### PR DESCRIPTION
**Pull request by leyshon**
## Description

This adds the IDToken to the session for the Azure provider.

## Motivation and Context

This allows it to be passed as a bearer token in the Authorization header to an upstream such as the Kubernetes dashboard. It solves the issue in https://github.com/pusher/oauth2_proxy/issues/131

## How Has This Been Tested?

This has been tested against Azure AD to check that it returns the IDToken. It was tested with nginx in auth proxy mode to check that this token could be used to authenticate to the kubernetes dashboard.

As a result of the change the cookie being returned is larger and has to be split. This can cause an issue if using nginx to proxy requiring proxy buffer size be increased (see https://andrewlock.net/fixing-nginx-upstream-sent-too-big-header-error-when-running-an-ingress-controller-in-kubernetes/). I'm not sure therefore if something should be noted in the changelog regarding this.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by leyshon** [comment]
I've added a note to both the documentation and the changelog

File: docs/2_auth.md [hunk 1]
```diff
@@ -80,6 +80,8 @@ Note: The user is checked against the group members list on initial authenticati
    --client-secret=<value from step 6>
 ```
 
+Note: When using the Azure Auth provider with nginx and the cookie session store you may find the cookie is too large and doesn't get passed through correctly. Increasing the proxy_buffer_size in nginx or implementing the redis session storage (see https://pusher.github.io/oauth2_proxy/configuration#redis-storage) should resolve this.
```
Comment (eerichmond) [inline docs/2_auth.md]: You might also mention that they might need to increase large_client_header_buffers to nginx

Comment (leyshon) [reply docs/2_auth.md]: I don't think large_client_header_buffers is needed for this change. The default in nginx for this is 8kb (https://nginx.org/en/docs/http/ngx_http_core_module.html#large_client_header_buffers) and the cookie resulting from this change is under 8kb. It's above 4kb though which is mostly the default for proxy_buffer_size (https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffer_size).

**Comment by iremmats** [comment]
We just tested this branch and ran into an issue with
[2019/10/03 06:23:31] [oauthproxy.go:745] Error loading cookied session: Cookie "_oa
uth2_proxy" not present

As you say in the conversation/documentation the cookie has been split into _oauth2_proxy_0, _oauth2_proxy_1, _oauth2_proxy_2. We are using the auth-request mode which seems to be the most common.

Nginx has been configured with increased proxy buffer size. Where do we need to make changes now to make this work?

This is how we configure our ingress:
nginx.ingress.kubernetes.io/auth-signin: https://$host/oauth2/start?rd=$request_uri
nginx.ingress.kubernetes.io/auth-url: https://$host/oauth2/auth

**Comment by leyshon** [comment]
@iremmats What do the logs for the ingress controller say? The log from oauth2-proxy says it's not getting the cookie and the reason for that lies somewhere in nginx so those logs should hold the answer.

**Comment by JoelSpeed** [comment]
@iremmats Split cookies with auth request mode can cause issues. Ensuring Nginx properly copies the set-cookie headers across has had many different solutions posted in issues on this repo for a while but it has not really been solved.

Instead, the recommendation is to use [redis for session storage](https://pusher.github.io/oauth2_proxy/configuration#redis-storage), which solves the problem of large cookies by just issuing a session ticket rather than the full session, meaning no more split cookies

File: docs/2_auth.md [hunk 2]
```diff
@@ -81,6 +81,8 @@ Note: The user is checked against the group members list on initial authenticati
...
```
Comment (JoelSpeed) [inline docs/2_auth.md]: Should be able to use a link in the markdown and have jekyll expand it, this should mean links work when testing locally too
```suggestion
Note: When using the Azure Auth provider with nginx and the cookie session store you may find the cookie is too large and doesn't get passed through correctly. Increasing the proxy_buffer_size in nginx or implementing the [redis session storage](configuration#redis-storage) should resolve this.
```

**Comment by xyxiaobai** [comment]
Any idea when this will be merged? I'm looking forward to testing it out! Thanks.

**Comment by syscll** [comment]
If we can get the conflicts resolved I think this will be good to go.

**Comment by leyshon** [comment]
@syscll I've resolved the conflicts if you're able to review it once more. Thanks.

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,8 @@
 - [#227](https://github.com/pusher/oauth2_proxy/pull/227) Add Keycloak provider (@Ofinka)
 - [#273](https://github.com/pusher/oauth2_proxy/pull/273) Support Go 1.13 (@dio)
 - [#275](https://github.com/pusher/oauth2_proxy/pull/275) docker: build from debian buster (@syscll)
+- [#258](https://github.com/pusher/oauth2_proxy/pull/258) Add IDToken for Azure provider
+  - This PR adds the IDToken into the session for the Azure provider allowing requests to a backend to be identified as a specific user. As a consequence, if you are using a cookie to store the session the cookie will now exceed the 4kb size limit and be split into multiple cookies. This can cause problems when using nginx as a proxy, resulting in no cookie being passed at all. Either increase the proxy_buffer_size in nginx or implement the redis session storage (see https://pusher.github.io/oauth2_proxy/configuration#redis-storage)
```
Comment (syscll) [inline CHANGELOG.md]: I feel like this note is definitely relevant but maybe the changelog is the wrong place? WDYT @JoelSpeed @steakunderscore?

Comment (JoelSpeed) [reply CHANGELOG.md]: I am happy to have a note like this in the changelog, it's important that people are aware of changes like this that may break their setups. 

It's the sort of thing I'll copy out to highlight it in the release highlights when I cut the next release

Comment (syscll) [reply CHANGELOG.md]: Totally happy for it to stay in the changelog :)

**Pull request merged**